### DISCUSSION TITLE
Dynamic learning rate

### DISCUSSION THREAD

**Discussion by ANAS1997AA** [ultralytics/ultralytics#8921] [Docs]
![Capture d'écran 2024-03-13 210851](https://github.com/ultralytics/ultralytics/assets/162803130/aada1b02-4046-408b-877b-89c6eca68c85)
Hello. I want to ask if I'm on the right way to apply the exponential decay to my training. Please can someone help me complete my script.

**Comment by jmarintur** [comment]
Yes, the `lr` you print at each epoch is the one you should get. 

As it is written in your code before calling the `scheduler.step()`, you'll need to:

1. Load your data and iterate through it. For every batch, you'll perform the forward pass, calculate the loss, run `optimizer.zero_grad()`, the backward pass (`loss.backward()`), and finally the `optimizer.step()`
2. Then, after finishing the epoch, run the `scheduler.step()`

Did you try to follow https://docs.ultralytics.com/modes/train/#introduction?

**Reply by ANAS1997AA** [reply]
Can you please show me how can I calculate the loss, run optimizer.zero_grad(), the backward pass (loss.backward()), and finally the optimizer.step() ? Thank you so much

**Reply by jmarintur** [reply]
Hi @ANAS1997AA, wouldn't it be easier to follow the link I shared with you? There, they show you how to prepare your data and train a model.

**Reply by ANAS1997AA** [reply]
Thank you so much

**Reply by ANAS1997AA** [reply]
i encountered this problem while running this line

 from ultralytics.datasets import COCODataset  

error : ModuleNotFoundError: No module named 'ultralytics.datasets'

**Reply by glenn-jocher** [reply]
Hi there! It looks like you're trying to import a module that doesn't exist in the Ultralytics YOLOv8 repository. For working with datasets in YOLOv8, you typically don't need to import datasets directly like this. Instead, you can specify your dataset in a YAML file and use it during training or validation with the YOLO model.

If you're looking to work directly with the COCO dataset for training, you might want to check out the dataset preparation guide in the [Ultralytics documentation](https://docs.ultralytics.com/). This will guide you on how to set up your dataset YAML file correctly.

For a quick example on how to specify your dataset for training:

```yaml
# coco128.yaml
train: ../datasets/coco128/images/train2017/  # 128 images
val: ../datasets/coco128/images/train2017/  # 128 images

# Classes
nc: 80  # number of classes
names: [ 'person', 'bicycle', 'car', ...]  # class names
```

And then you can train your model using this dataset by specifying the YAML file:

```bash
yolo train --data coco128.yaml --cfg yolov8n.yaml
```

Hope this helps! If you have any more questions, feel free to ask. 😊
